### ISSUE TITLE
Not OK: perl 5.00561 on PA-RISC1.1 10.20

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#651]
Migrated from [rt.perl.org#1527](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1527) (status was 'resolved')

Searchable as RT1527$

**Comment by p5pRT** [comment]
### From lewart@www.cvm.uiuc.edu




---
Perl5\-Porters\,

getpwent and getpwuid are broken\, but getpwnam is okay&#8203;:
1\) getpwent returns the last entry in /etc/passwd \.
2\) getpwuid\($num\) returns the last entry in /etc/passwd if $num is
&nbsp;   a valid UID; otherwise it returns an empty string\.

Why didn't "make test" catch such egregious errors?  Because of the
following stanza from t/op/pwent\.t&#8203;:
&nbsp;  94&#8203;: \# Protect against one\-to\-many and many\-to\-one mappings\.
&nbsp;  95&#8203;: if \($name\_s ne $name\) \{
&nbsp;  96&#8203;:     @&#8203;n = getpwnam\($name\_s\);
&nbsp;  97&#8203;:     \($name\,$passwd\,$uid\,$gid\,$quota\,$comment\,$gcos\,$home\,$shell\) = @&#8203;n;
&nbsp;  98&#8203;:     next if $name\_s ne $name;
&nbsp;  99&#8203;: \}
What is the point of the above clause?

Cheers\,
Daniel Lewart
d\-lewart@&#8203;uiuc\.edu


---


---
Site configuration information for perl 5\.00561&#8203;:

Configured by lewart at Mon Sep  6 14&#8203;:01&#8203;:38 CDT 1999\.

Summary of my perl5 \(revision 5\.0 version 5 subversion 61\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=hpux\, osvers=10\.20\, archname=PA\-RISC1\.1
&nbsp;    uname='hp\-ux www b\.10\.20 a 9000715 2013547832 two\-user license '
&nbsp;    config\_args='\-de \-Dcc=gcc \-Dprefix=/ext/perl5\.005\_61'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef useperlio=undef d\_sfio=undef
&nbsp;    use64bits=undef usemultiplicity=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, optimize='\-O'\, gccversion=2\.95\.1 19990816 \(release\)
&nbsp;    cppflags='\-D\_HPUX\_SOURCE \-I/usr/local/include'
&nbsp;    ccflags ='\-D\_HPUX\_SOURCE \-I/usr/local/include'
&nbsp;    stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=8
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='ld'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib/pa1\.1 /lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lnsl\_s \-lndbm \-lgdbm \-ldb \-ldld \-lm \-lc \-lndir \-lcrypt \-lsec
&nbsp;    libc=/lib/libc\.sl\, so=sl\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_hpux\.xs\, dlext=sl\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E \-Wl\,\-B\,deferred '
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-b \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.00561&#8203;:
&nbsp;    /ext/perl5\.005\_61/lib/5\.00561/PA\-RISC1\.1
&nbsp;    /ext/perl5\.005\_61/lib/5\.00561
&nbsp;    /ext/perl5\.005\_61/lib/site\_perl/5\.00561/PA\-RISC1\.1
&nbsp;    /ext/perl5\.005\_61/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.00561&#8203;:
&nbsp;    HOME=/home/lewart
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/ccs/bin&#8203;:/usr/contrib/bin&#8203;:/opt/nettladm/bin&#8203;:/opt/pd/bin&#8203;:/usr/bin/X11&#8203;:/usr/contrib/bin/X11&#8203;:/opt/upgrade/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/bin/ksh
&nbsp;    SHLIB\_PATH \(unset\)

**Comment by p5pRT** [comment]
### From @jhi



Daniel S\. Lewart writes&#8203;:
 > getpwent and getpwuid are broken\, but getpwnam is okay&#8203;:
 > 1\) getpwent returns the last entry in /etc/passwd \.
 > 2\) getpwuid\($num\) returns the last entry in /etc/passwd if $num is
 >    a valid UID; otherwise it returns an empty string\.

Huh?  This sounds so weird that I have to check&#8203;: getpwent\(\) \*always\*
returns only the last entry in /etc/passwd\(\)?  After
endpwent\(\)\+setpwent\(\)\, too?

Ditto for getpwuid\($valid\_num\)\.

The "empty string" case of getpwuid\($invalid\_num\) is as it should be\.
If you look closer you'll notice it's the undef \(or an empty list
in array context\)\.

\(I checked almost\-\_62 in Digital UNIX and getpwent\(\) and getpwuid\(\)
 work okay\.  I'll try compiling the same Perl tomorrow\-\-but I don't
 have HP\-UX 10\.20 anymore\, I have 11\.0\, but I don't have gcc in there\,
 only the HP ANSI cc\.\)

 > Why didn't "make test" catch such egregious errors?  Because of the
 > following stanza from t/op/pwent\.t&#8203;:
 >   94&#8203;: \# Protect against one\-to\-many and many\-to\-one mappings\.
 >   95&#8203;: if \($name\_s ne $name\) \{
 >   96&#8203;:     @&#8203;n = getpwnam\($name\_s\);
 >   97&#8203;:     \($name\,$passwd\,$uid\,$gid\,$quota\,$comment\,$gcos\,$home\,$shell\) = @&#8203;n;
 >   98&#8203;:     next if $name\_s ne $name;
 >   99&#8203;: \}
 > What is the point of the above clause?

The passwd file might have both one username \-> many uids mappings
\(illegal but possible\) and one uid \-> many usernames mappings \(slightly
less illegal\, but sometimes useful\)\.  We cannot do many \(any?\) useful
checks if we have any many\-mappings\.

I must be slow but could you elaborate on how the above test makes
getpw\*\(\) tests to fail \(to give a false positive\)?

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From @jhi


Daniel S\. Lewart writes&#8203;:
 > 
 > This is a build failure report for perl from d\-lewart@&#8203;uiuc\.edu\,
 > generated with the help of perlbug 1\.27 running under perl 5\.00561\.
 > 
 > 
 > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
 > Perl5\-Porters\,
 > 
 > getpwent and getpwuid are broken\, but getpwnam is okay&#8203;:
 > 1\) getpwent returns the last entry in /etc/passwd \.
 > 2\) getpwuid\($num\) returns the last entry in /etc/passwd if $num is
 >    a valid UID; otherwise it returns an empty string\.

Soon\-to\-be\-62 in HP\-UX 11\.0 compiled with gcc 2\.8\.1&#8203;: getpwent\(\) most
certainly doesn't return the last entry in /etc/passwd \-\- not until
the last one is supposed to be returned\.
getpwuid\($valid\_num\) works okay\, too\.

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen
